### PR TITLE
Action parsing

### PR DESCRIPTION
**Pull request by dcolestock**
@jswaro: Want to take a look? I have a number of major revisions and additions

### CODE REVIEW

**Comment by dcolestock** [comment]
@jswaro: You'll notice I finally got around to adding some unit tests.  Right now they are a bit too syntax specific as they require the bot to respond with very specifically worded responses, so I plan to make it more result based as we move on.

Also, I notice travis-ci is only running on python 3.2-3.5, but I thought you wanted coupbot to remain 2.7 compatible, so thought it might be worth seeing if you can add more versions to that.

File: gamebot/coup/actions.py [hunk 1]
```diff
@@ -11,245 +10,384 @@
...
 @action_register(do_action)
-class Tax(BaseAction):
+class Tax(CoupAction):
     name = "Tax"
     description = "Take 3 coins"
 
     @staticmethod
-    def do_success(target, source):
-        source.modify_coins_by_action(3)
+    def run(game, source_player):
```
Comment (jswaro) [inline gamebot/coup/actions.py]: Why 'run'? Why is this necessary?

File: gamebot/coup/actions.py [hunk 2]
```diff
@@ -11,245 +10,384 @@
...
+        if target_player == source_player:
+            raise GameInvalidOperation("Cannot assassinate from yourself. Choose a different action")
+        if target_player.dead():
+            raise GameInvalidOperation("{} is dead. Choose a different action".format(target_player))
+        if same_team(source_player, target_player):
+            raise GameInvalidOperation("{} is on your team. Choose a different action".format(target_player))
+        source_player.modify_coins(-3)
+        game.broadcast_message("{} assassinates {}".format(source_player, target_player))
+        if len(target_player.available_influence) > 1:
+            game.add_message_to_queue(target_player.name, "Please challenge, counter, or select which card to "
```
Comment (jswaro) [inline gamebot/coup/actions.py]: Copy paste error?

File: gamebot/coup/exceptions.py [hunk 3]
```diff
@@ -1,4 +1,3 @@
-__author__ = "jswaro"
```
Comment (jswaro) [inline gamebot/coup/exceptions.py]: We should leave the docstring variables in files that are actually in use.

File: gamebot/game/team.py [hunk 4]
```diff
@@ -3,5 +3,5 @@ class BaseTeam(object):
 
 
 def same_team(target, source):
-    return target.team == source.team and target.team != BaseTeam.UNASSIGNED
+    return target.team == source.team and target.team != BaseTeam.UNASSIGNED  # TODO: Handle all same team
```
Comment (jswaro) [inline gamebot/game/team.py]: This todo should be present in the redefined team for coup. This type of logic is fine for the general case, but doesn't handle the way coup might behave.

File: gamebot/coup/actions.py (hunk 1 above)
Comment (dcolestock) [inline gamebot/coup/actions.py]: Run is what processes and announces the initial action adding it to the stack opening it up for potential responses.  If you had another idea of how to implement this part I'm open to doing it a different way.

File: gamebot/coup/actions.py (hunk 2 above)
Comment (dcolestock) [inline gamebot/coup/actions.py]: Which error?

File: gamebot/coup/exceptions.py (hunk 3 above)
Comment (dcolestock) [inline gamebot/coup/exceptions.py]: The __author__ docstrings weren't consistently in every file and where they were I could've added myself, but don't see the benefit.

http://stackoverflow.com/questions/1523427/what-is-the-common-header-format-of-python-files#comment11616749_1523456

> I think all of this metadata after the imports is a bad idea. The parts of this metadata that apply to a single file (eg author, date) are already tracked by source control. Putting an erroneous & out of date copy of the same info in the file itself seems wrong to me. The parts that apply to the whole project (eg licence, versioning) seem better located at a project level, in a file of their own, rather than in every source code file.

Let me know if you don't agree with this approach or reasoning.

**Comment by jswaro** [comment]
:+1:
